builtin/providers/azure/resource_azure_disk.go

```diff
@@ -0,0 +1,301 @@
+package azure
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/svanharmelen/azure-sdk-for-go/management"
+	"github.com/svanharmelen/azure-sdk-for-go/management/virtualmachinedisk"
+)
+
+const dataDiskBlobStorageURL = "http://%s.blob.core.windows.net/disks/%s.vhd"
+
+func resourceAzureDataDisk() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAzureDataDiskCreate,
+		Read:   resourceAzureDataDiskRead,
+		Update: resourceAzureDataDiskUpdate,
+		Delete: resourceAzureDataDiskDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+
+			"lun": &schema.Schema{
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+
+			"size": &schema.Schema{
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+
+			"caching": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "None",
+			},
+
+			"storage": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"media_link": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+
+			"source_media_link": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"virtual_machine": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+		},
+	}
+}
+
+func resourceAzureDataDiskCreate(d *schema.ResourceData, meta interface{}) error {
+	mc := meta.(management.Client)
+
+	if err := verifyDataDiskParameters(d); err != nil {
+		return err
+	}
+
+	lun := d.Get("lun").(int)
+	vm := d.Get("virtual_machine").(string)
+
+	p := virtualmachinedisk.CreateDataDiskParameters{
+		Lun:                 lun,
+		LogicalDiskSizeInGB: d.Get("size").(int),
+		HostCaching:         hostCaching(d),
+		MediaLink:           mediaLink(d),
+		SourceMediaLink:     d.Get("source_media_link").(string),
+	}
+
+	if name, ok := d.GetOk("name"); ok {
+		p.DiskName = name.(string)
+	}
+
+	log.Printf("[DEBUG] Adding data disk %d to instance: %s", lun, vm)
+	req, err := virtualmachinedisk.NewClient(mc).AddDataDisk(vm, vm, vm, p)
```

Comment:
This is troublesome when you consider instances and hosted services as decoupled and independent resources.
I'd go ahead and add an optional service_name parameter and use that if it's given; else default to the vm's name.